docs/source/ckpt_converters/dev_guide.rst

```diff
@@ -0,0 +1,218 @@
+Community Model Converter Development Guide
+===========================================
+
+Guideline Steps for Checkpoint Conversion
+-----------------------------------------
+
+1. **Understand Both Frameworks**: Familiarize yourself with the architectures and naming conventions of both HuggingFace and NeMo models.
+
+2. **Load Community Checkpoint**: For example, use HuggingFace’s ``AutoModel`` to load the pre-trained model.
+
+3. **Inspect Model and Config**: Understand the layer names, parameter shapes, and essential configs.
+
+4. **Adjust NeMo Model Configuration**: Modify the NeMo model configuration to match the HuggingFace model’s specifications.
+
+5. **Initialize NeMo Model**: Create an instance of the corresponding NeMo model.
+
+6. **Create Key Mapping**: Define a function to map HuggingFace layer names to NeMo layer names. Adjust for any structural differences.
+
+7. **Rename and Reshape Parameters**: Implement a function to rename keys in the HuggingFace state dictionary and reshape tensors if necessary. For example, QKV weights usually need some special handling from HF to NeMo.
+
+8. **Load Converted Weights into NeMo Model**: Apply the transformed state dictionary to the NeMo model.
+
+9. **Save NeMo Checkpoint**: Save the updated NeMo model as a new checkpoint.
+
+10. **Verification**: Verify the performance of the NeMo model to ensure successful conversion.
+
+11. **Add Docstrings and Comments**: Please kindly comment the expected shapes in the parameter reshaping part.
+
```

Comment:
Somewhere here: bullet point to setup Jenkins test (possibly for a tiny model).

This function can be used to create a model programmatically https://github.com/NVIDIA/NeMo/blob/main/tests/setup/models/create_hf_model.py (for reproducibility). It can be called like this, see [here](https://github.com/NVIDIA/NeMo/blob/48b8204d57e59c8790aaa6eaa20384b046b1a574/tests/setup/__main__.py#L36-L41):
```
create_hf_model(
    model_name_or_path="/home/TestData/nlp/meta-llama/Llama-2-7b-hf",
    output_dir=os.path.join(args.save_dir, "megatron_llama/llama-ci-hf"),
    config_updates={"hidden_size": 256, "num_attention_heads": 4, "num_hidden_layers": 2, "num_key_value_heads": 4},
    overwrite=args.overwrite,
)
```

Reply:
Added jenkins. For HF model creation, I added to the tips.